### PR TITLE
Revision of BYOL module and tests

### PR DESCRIPTION
**Pull request by matsumotosan**
## What does this PR do?

Fixes part of #839

- `pl_bolts.models.self_supervised.byol.byol_module.BYOL`
- `pl_bolts.models.self_supervised.byol.models.MLP`
- `pl_bolts.models.self_supervised.byol.models.SiameseArm`
- `pl_bolts.tests.models.self_supervised.test_models.test_byol`

I cannot fix the CLI script as I do not have a GPU
`pl_bolts.tests.models.self_supervised.test_models.test_ssl_scripts.test_cli_run_ssl_byol`

## Summary

- [x] Add method in `BYOL` to calculate loss for given online and target views
- [x] Add initial value of tau as an argument for `BYOL`
- [x] Remove predictor network from `SiameseArm` since it was not being used for target network and move to `BYOL`
- [x] Modify `test_byol` to use SimCLR transforms instead of CPC transforms
- [x] Fix CLI run test case

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does **only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? \[not needed for typos/docs\]
- [x] Did you verify new and **existing tests** pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review

- [ ] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 1]
```diff
@@ -1,35 +1,39 @@
...
-@under_review()
 class BYOL(LightningModule):
     """PyTorch Lightning implementation of Bootstrap Your Own Latent (BYOL_)_
 
     Paper authors: Jean-Bastien Grill, Florian Strub, Florent Altché, Corentin Tallec, Pierre H. Richemond, \
     Elena Buchatskaya, Carl Doersch, Bernardo Avila Pires, Zhaohan Daniel Guo, Mohammad Gheshlaghi Azar, \
     Bilal Piot, Koray Kavukcuoglu, Rémi Munos, Michal Valko.
 
-    Model implemented by:
-        - `Annika Brundyn <https://github.com/annikabrundyn>`_
```
Comment (otaj) [inline pl_bolts/models/self_supervised/byol/byol_module.py:25]: We don't want to throw away attribution

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 2]
```diff
@@ -214,13 +199,16 @@ def cli_main():
...
         (c, h, w) = dm.dims
         dm.train_transforms = SimCLRTrainDataTransform(h)
         dm.val_transforms = SimCLREvalDataTransform(h)
         args.num_classes = dm.num_classes
+    else:
+        raise ValueError(
+            f"{args.dataset} is not a valid dataset. Dataset must be 'cifar10', 'stl10', or 'imagenet2012'."
+        )
 
     model = BYOL(**args.__dict__)
```
Comment (otaj) [inline pl_bolts/models/self_supervised/byol/byol_module.py:213]: I know this was already here, but it looks quite hacky. Can we atleast do something like this?

```suggestion
    model = BYOL(**vars(args))
```

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 3]
```diff
@@ -75,77 +74,70 @@ def __init__(
...
-            num_workers: number of workers
-            warmup_epochs: num of epochs for scheduler warm up
-            max_epochs: max epochs for scheduler
-            base_encoder: the base encoder module or resnet name
-            encoder_out_dim: output dimension of base_encoder
-            projector_hidden_size: hidden layer size of projector MLP
-            projector_out_dim: output size of projector MLP
-        """
+        initial_tau: float = 0.996,
+        **kwargs: Any,
```
Comment (otaj) [inline pl_bolts/models/self_supervised/byol/byol_module.py:80]: Quite a few of these args is not used, can we get rid of them? (I cannot comment on them specifically since you didn't change the code around them)

File: tests/models/self_supervised/test_models.py [hunk 4]
```diff
@@ -37,16 +35,15 @@ def test_cpcv2(tmpdir, datadir):
     trainer.fit(model, datamodule=datamodule)
 
 
-# todo: some pickling issue with min config
-@pytest.mark.skipif(LooseVersion(torch.__version__) < LooseVersion("1.7.0"), reason="Pickling issue")
 def test_byol(tmpdir, datadir):
```
Comment (otaj) [inline tests/models/self_supervised/test_models.py:38]: Can we add `catch_warnings` fixture in here?

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 5]
```diff
@@ -75,77 +74,70 @@ def __init__(
...
+        loss_12, loss_21, total_loss = self._shared_step(batch, batch_idx)
+        self.log_dict({"loss_12": loss_12, "loss_21": loss_21, "val_loss": total_loss})
         return total_loss
 
-    def validation_step(self, batch, batch_idx):
-        loss_a, loss_b, total_loss = self.shared_step(batch, batch_idx)
+    def _shared_step(self, batch: Any, batch_idx: int) -> Tuple[Tensor, Tensor, Tensor]:
+        """Shared evaluation step for training and validation loop."""
+        imgs, _ = batch
+        img1, img2 = imgs[:2]
```
Comment (otaj) [inline pl_bolts/models/self_supervised/byol/byol_module.py:131]: Do I read it correctly that it actually uses only first two images from the batch and discards the rest? Surely that can't be right

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 6]
```diff
@@ -75,77 +74,70 @@ def __init__(
...
-        self.log_dict({"1_2_loss": loss_a, "2_1_loss": loss_b, "val_loss": total_loss})
+        # Calculate similarity loss in each direction
+        loss_12 = self.calculate_loss(img1, img2)
+        loss_21 = self.calculate_loss(img2, img1)
 
-        return total_loss
+        # Calculate total loss
+        total_loss = loss_12 + loss_21
+
+        return loss_12, loss_21, total_loss
```
Comment (otaj) [inline pl_bolts/models/self_supervised/byol/byol_module.py:140]: Can the shared step also include logging? Although `training_step` and `validation_step` are only three lines of code, essentially all of it is the same except one string, so it could really be only one line of code with a parameter

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 7]
```diff
@@ -155,30 +147,29 @@ def configure_optimizers(self):
...
         (args, _) = parser.parse_known_args()
 
-        # Data
+        # Data params
         parser.add_argument("--data_dir", type=str, default=".")
         parser.add_argument("--num_workers", default=8, type=int)
 
-        # optim
+        # Training params
         parser.add_argument("--batch_size", type=int, default=256)
```
Comment (otaj) [inline pl_bolts/models/self_supervised/byol/byol_module.py:162]: Same as previous line

**Review by otaj — COMMENTED** [review]
Hi, @matsumotosan! Thank you for this PR. I have to say, I love what you've done with `byol_module` in general, the meat of this module is now much more readable.

However, I left couple comments, which are mostly related to passing arguments around (similar to my comments in #873)

File: pl_bolts/models/self_supervised/byol/byol_module.py (hunk 3 above)
Comment (matsumotosan) [reply pl_bolts/models/self_supervised/byol/byol_module.py:80]: I am keeping `learning_rate`, `weight_decay`, `warmup_epochs`, and `max_epochs` since they are called from `self.hparams` in `configure_optimizers()`.

File: pl_bolts/models/self_supervised/byol/byol_module.py (hunk 5 above)
Comment (matsumotosan) [reply pl_bolts/models/self_supervised/byol/byol_module.py:131]: The data are loaded using `SimCLRTrainDataTransform` so `imgs` is a 3-tuple with two transforms and one online transform of `batch_size` images. The online transform is discarded for training and `img1` and `img2` are each (B, C, H, W) tensors.

Comment (otaj) [reply pl_bolts/models/self_supervised/byol/byol_module.py:131]: Oh, absolutely, I somehow missed that, but you are correct. Thank you for clarification! :zap:

File: pl_bolts/models/self_supervised/byol/byol_module.py [hunk 8]
```diff
@@ -155,30 +147,29 @@ def configure_optimizers(self):
...
         parser = ArgumentParser(parents=[parent_parser], add_help=False)
         parser.add_argument("--online_ft", action="store_true", help="run online finetuner")
         parser.add_argument("--dataset", type=str, default="cifar10", choices=["cifar10", "imagenet2012", "stl10"])
 
         (args, _) = parser.parse_known_args()
 
-        # Data
+        # Data params
         parser.add_argument("--data_dir", type=str, default=".")
         parser.add_argument("--num_workers", default=8, type=int)
```
[the remark this answers is not in the document]
Comment (otaj) [reply pl_bolts/models/self_supervised/byol/byol_module.py:159]: I am fine with either. I think it might be useful to have these `add_dataset_specific_args` methods on datasets. If you decide to go with that route, then such PR would have to be merged before this one (and #873)

File: pl_bolts/models/self_supervised/byol/byol_module.py (hunk 3 above)
Comment (otaj) [reply pl_bolts/models/self_supervised/byol/byol_module.py:80]: I don't think that `max_epochs` is used in `self.hparams` (and even if it is, it is not really appropriate, since it's not model-specific, but trainer-specific argument)

File: pl_bolts/models/self_supervised/byol/byol_module.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (matsumotosan) [reply pl_bolts/models/self_supervised/byol/byol_module.py:159]: I will add a `add_dataset_specific_args` method to the CIFAR10, STL10, ImageNet2012, and MNIST datamodules. I will try to open a PR in the next couple of days.

File: pl_bolts/models/self_supervised/byol/byol_module.py (hunk 3 above)
Comment (matsumotosan) [reply pl_bolts/models/self_supervised/byol/byol_module.py:80]: Is it not necessary for configuring the scheduler (LinearWarmupCosineAnnealingLR)?

Comment (otaj) [reply pl_bolts/models/self_supervised/byol/byol_module.py:80]: I feel like I'm overlooking too much lately :facepalm: Of course it is used in the scheduler and therefore very much appropriate in here.

**Review by otaj — APPROVED** [review]

**Comment by matsumotosan** [comment]
@otaj I realized that the BYOL module is initialized with `data_dir=datadir` in `test_byol`. Should I re-open this PR to fix it or create a new PR?
